sparrow_datums/keypoints/keypoints.py

```diff
@@ -0,0 +1,197 @@
+    >>> import numpy as np
+    >>> from sparrow_datums import Keypoints, PType
+    >>> Keypoints(np.ones(2))
+    Keypoints([1., 1.])
+    """
+
+    def validate(self) -> None:
+        """Check validity of boxes array."""
+        if not self.shape or self.shape[-1] != 2:
+            raise ValidationError("Keypoint arrays must have size-2 dimensions")
+        elif self.ndim > 2:
+            raise ValidationError(
+                "2D Keypoint chunks are the maximum dimension supported."
+            )
+
+    @property
+    def scale(self) -> FloatArray:
+        """Scaling array."""
+        if self._scale is None:
+            width = self.image_width
+            height = self.image_height
+            self._scale = np.array([width, height])
+        return self._scale
+
+    @property
+    def is_relative(self) -> bool:
+        """Parameterization is relative."""
+        return bool(self.ptype.is_relative)
+
+    @property
+    def is_absolute(self) -> bool:
+        """Parameterization is absolute."""
+        return bool(self.ptype.is_absolute)
+
+    def to_relative(self: T) -> T:
+        """Convert keypoint to relative pixel coordinates, if necessary."""
+        if self.is_relative:
+            return self
+        x = self.array.copy().astype(np.float32)
+        x[..., :2] /= self.scale
+        if self.ptype is PType.unknown:
+            param_type = PType.relative_xy
+        else:
+            param_type = self.ptype.as_relative
+        return self.__class__(
+            x,
+            ptype=param_type,
+            **self.metadata_kwargs,
+        )
+
+    def to_absolute(self: T) -> T:
+        """Convert keypoint to absolute pixel coordinates, if necessary."""
+        if self.is_absolute:
+            return self
+        x = self.array.copy().astype(np.float32)
+        x[..., :2] *= self.scale
+        if self.ptype is PType.unknown:
+            param_type = PType.absolute_xy
+        else:
+            param_type = self.ptype.as_absolute
+        return self.__class__(
+            x,
+            ptype=param_type,
+            **self.metadata_kwargs,
+        )
+
+    def validate_known_ptype(self) -> None:
+        """Make sure PType is a known box parameterization."""
+        known_box_parameterizations = {PType.absolute_xy, PType.relative_xy}
+        if self.ptype not in known_box_parameterizations:
+            raise ValidationError(f"Unknown box parameterization: {self.ptype.name}")
+
+    @property
+    def x(self) -> FloatArray:
+        """Slice the x dimension of the keypoint."""
+        self.validate_known_ptype()
+        result: FloatArray
+        result = self.array[..., 0]
+        return result
+
+    @property
+    def y(self) -> FloatArray:
+        """Slice the y dimension of the keypoint."""
+        self.validate_known_ptype()
+        result: FloatArray
+        result = self.array[..., 1]
+        return result
+
+    def __generate_heatmap(self, x0: int, y0: int, covariance: float):
```

Comment:
Let's just use a single underscore here. I think that's the preferred pattern:

```suggestion
    def _generate_heatmap(self, x0: int, y0: int, covariance: float):
```

Reply:
done